### DISCUSSION TITLE
OpenAI API option Tool_choice: "required"

### DISCUSSION THREAD

**Discussion by jimmystridh** [microsoft/semantic-kernel#6352] [Q&A] [answered]
From the OpenAI api docs:
> To force the model to always call one or more functions, you can set tool_choice: "required". The model will then select which function(s) to call.

This does not seem to be abstracted or exposed in any way as far as I can find.

We'd like to give the model a number of tools and require it use one of them rather than just give a response, and then this seems useful.

Is it possible to use this setting, or thoughts on adding support for it?

Thanks

**Comment by stephentoub** [comment, marked as answer]
That support was only very recently added to OpenAI. ToolCallBehavior currently has a RequireFunction method that accepts a single function:
https://github.com/microsoft/semantic-kernel/blob/2b96abfb2fb5d3cc3295cd9aec1ef689befb722d/dotnet/src/Connectors/Connectors.OpenAI/ToolCallBehavior.cs#L81
It should be trivial to add an overload that accepts an enumerable of functions. There is also work underway to abstract all of this away in order to allow it all to be used in a service-agnostic way, and that abstraction already factors in this capability.

**Reply by stephentoub** [reply]
cc: @SergeyMenshykh

**Comment by SergeyMenshykh** [comment]
Yep, the new abstraction will accept many functions as required ones. The Azure.AI.OpenAI SDK, which SK uses to talk to the {Azure}OpenAI service, should add support for that as well before we have end-to-end support for multiple required functions.

**Reply by jimmystridh** [reply]
Thanks! Even if it does sound a bit cumbersome to do "use a plugin, any of the provided ones to the kernel" as we'd have to iterate over the plugins and provide those as required functions, that would technically solve our problem.
